### PR TITLE
Create AWS test accounts using terraform

### PR DESCRIPTION
**Pull request by justinsb**
This brings them under declarative management.  We use a date to
enable rotation of old accounts, as these are limited.  Pre-created 14
accounts, next step is to get these into boskos.

A lot of workarounds for terraform shortcomings, hence the slightly
odd structure.

### CODE REVIEW

**Comment by justinsb** [comment]
cc @BenTheElder @detiber 

I also have a bunch of yaml files ready to upload into the boskos cluster!

**Comment by BenTheElder** [comment]
seems about right, I'm not 100% positive about this though.
go.k8s.io/oncall has the current test-infra oncall for uploading to boskos

**Comment by detiber** [comment]
/lgtm
I'm not overly familiar with terraform, but nothing is jumping out at me.

**Comment by justinsb** [comment]
Storing it in a GCS bucket seems reasonable.  The tfstate file does include AWS account credentials (for now), so we may want to look into encryption.

This must be run by people with administrative access to the AWS CNCF account.  I suggest we aim to follow in the footsteps of the GCP automation - is that still run manually?

The other problem is that we must upload the resulting credentials to boskos (when we create new accounts).  I don't believe we've done that yet here.

In the long term it would be nice to get the prow jobs using workload identity to authenticate to AWS, so there would be no need to store credentials.  But that feels like it's going to require some test-infra changes.

**Comment by thockin** [comment]
On Wed, Aug 12, 2020 at 12:38 PM Justin Santa Barbara <
notifications@github.com> wrote:

> Storing it in a GCS bucket seems reasonable. The tfstate file does include
> AWS account credentials (for now), so we may want to look into encryption.
>
> This must be run by people with administrative access to the AWS CNCF
> account. I suggest we aim to follow in the footsteps of the GCP automation
> - is that still run manually?
>

Yes but the set of people who can run it is governed by GoogleGroups.


> The other problem is that we must upload the resulting credentials to
> boskos (when we create new accounts). I don't believe we've done that yet
> here.
>
> In the long term it would be nice to get the prow jobs using workload
> identity to authenticate to AWS, so there would be no need to store
> credentials. But that feels like it's going to require some test-infra
> changes.
>

Agree


>

File: infra/aws/e2e/accounts/tf/modules/user/main.tf
Comment (randomvariable) [inline infra/aws/e2e/accounts/tf/modules/user/main.tf:34]: Also add `AmazonSSMFullAccess` as per https://github.com/kubernetes/k8s.io/issues/984

Comment (justinsb) [reply infra/aws/e2e/accounts/tf/modules/user/main.tf:34]: Done!

**Comment by justinsb** [comment]
Updated:

* Now stores tfstate encrypted in GCS.
* Also stores the yaml to apply to boskos in GCS, encrypted with the same key, for easier application to the cluster.
* README and scripts to simplify the process

File: infra/aws/e2e/accounts/tf/accounts.tf
Comment (onlydole) [inline infra/aws/e2e/accounts/tf/accounts.tf:15]: Style Nit: For these module instantiations, you might be able to [take advantage of the `for_each` command](https://www.terraform.io/docs/configuration/meta-arguments/for_each.html) and move these to be variables.

**Review by onlydole — CHANGES_REQUESTED** [review]
One small nit and I'm free to sync on that front if beneficial!

File: infra/aws/e2e/accounts/tf/accounts.tf
Comment (spiffxp) [inline infra/aws/e2e/accounts/tf/accounts.tf:2]: All of our other terraform modules require a minimum of 0.12 at present; can this be downgraded to 0.12?

File: infra/aws/e2e/accounts/tf/README.md
Comment (spiffxp) [inline infra/aws/e2e/accounts/tf/README.md:16]: Who has access to this?

Ideally this key would be stored in google cloud secret manager, and access would be restricted to a @kubernetes.io group we manage.  e.g.

```sh
KEY=$(gcloud secrets --project=kubernetes-public versions access latest --secret=gcs-encryption-key)
```

File: infra/aws/e2e/accounts/tf/README.md
Comment (spiffxp) [inline infra/aws/e2e/accounts/tf/README.md:21]: Who has access to this?

File: infra/aws/e2e/accounts/tf/modules/account-and-user/main.tf
Comment (spiffxp) [inline infra/aws/e2e/accounts/tf/modules/account-and-user/main.tf:1]: I'm not sure I understand what problem this is solving.  Why not merge the account and user modules into this one module?

File: infra/aws/e2e/accounts/tf/README.md
Comment (spiffxp) [inline infra/aws/e2e/accounts/tf/README.md:33]: Which cluster/boskos are these destined for?

Would like input from @kubernetes/test-infra-admins (as proxy for test-infra-oncall)

File: infra/aws/e2e/accounts/tf/tf-apply
Comment (spiffxp) [inline infra/aws/e2e/accounts/tf/tf-apply:15]: IMO it'd be better to store this directly in secret manager as well, then apply via e.g.

```
gcloud secrets --project=kubernetes-public versions access latest --secret=boskos-aws-accounts |\
    kubectl apply -f -
```

File: infra/aws/e2e/accounts/tf/accounts.tf
Comment (spiffxp) [inline infra/aws/e2e/accounts/tf/accounts.tf:117]: Do these need to be addressed?

**Review by spiffxp — CHANGES_REQUESTED** [review]
I want to unblock, but:
- I don't have appropriate access to run all of this
- My terraform environment isn't setup for this (I'll look into whether upgrading to 0.13 or 0.14 for all of our other modules is a heavy lift)
- I would rather see more use of `gcloud secrets` vs. storing encrypted in GCS

I see gs://k8s-infra-clusters-terraform/aws/k8s/accounts-boskos.yaml already exists, so if we want to unblock, I'm assuming it's a matter of getting test-infra-oncall access to that file and manually uploading?

File: infra/aws/e2e/accounts/tf/README.md
Comment (cjwagner) [reply infra/aws/e2e/accounts/tf/README.md:33]: This would need to be destined for the boskos instance in the build cluster where the consuming jobs will be run in order for the jobs to access the boskos instance. Ideally this would be a build cluster besides the default or trusted clusters so that test-infra-oncall doesn't need to act as a middleman for uploading to boskos or debugging related problems.

**Comment by spiffxp** [comment]
/close
Closing this in favor of https://github.com/kubernetes/k8s.io/pull/1839

**Pull request closed without merging**
